File: leetcode/0039/combinationSum.py
from typing import List
def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
    n = len(candidates)
    ans = []
    candidates.sort()
    def search(target,cur,begin):
        if target == 0:
            ans.append(cur[:])
        else:
            for i in range(begin,n):
                if target-candidates[i] < 0:
                    break
                cur.append(candidates[i])
                search(target-candidates[i],cur,i)
                cur.pop()
    search(target,[],0)
    return ans

File: leetcode/0039/combinationSum.swift
var n = 0
var ans = [[Int]]()
var sortedCandidates = [Int]()
func combinationSum(_ candidates: [Int], _ target: Int) -> [[Int]] {
    n = candidates.count
    ans = [[Int]]()
    sortedCandidates = candidates.sorted()
    
    dfs(target, [], 0)
    
    return ans
}

func dfs(_ target: Int, _ cur: [Int], _ begin: Int) {
    var cur = cur
    if target == 0 {
        ans.append(cur)
    } else {
        for i in begin..<n {
            if target - sortedCandidates[i] < 0 {
                break
            }
            cur.append(sortedCandidates[i])
            dfs(target-sortedCandidates[i], cur, i)
            cur.popLast()
        }
    }
}

print(combinationSum([2,3,6,7], 7))
